rfcs/min-relations.md

```diff
@@ -0,0 +1,438 @@
+Just like components, they can store effectively arbitrary data, or they can be data-less unit structs that serve as **markers**. Unlike components, they don't store the target entity directly, as that connection is managed by the engine.
+
+Each relation has three parts:
+
+1. The **source entity** that it is stored on.
+2. The **target entity** (or simply **target**) that it points to.
+3. The **relation kind** that defines the data that is stored on it.
+
+Making your first connection is easy, and mirrors the API that you're familiar with from components:
+
+```rust
+// Relations might just be markers
+struct FriendsWith;
+
+// But they might store data instead
+struct Owes(i32);
+
+// Let's make some entities
+let world = World::new();
+let alice = world.spawn().id();
+let boxy = world.spawn().id();
+
+// You can make relations with commands!
+commands.entity(alice).insert_relation(FriendsWith, boxy); // uwu:3
+
+// You can remove them with commands!
+commands.entity(alice).remove_relation(FriendsWith, boxy) // not uwu :(
+
+// You can use direct World access!
+world.entity_mut(alice).insert_relation(FriendsWith, boxy); // uwu:3
+
+// Relations are directed; a relation in one direction does not imply the same in the other direction
+commands.entity(boxy).insert_relation(FriendsWith, alice); // one way friendship sad :(
+
+// You can mix different relation kinds on the same entity!
+commands.entity(alice).insert_relation(Owes(9999), boxy); // :)))))
+
+// You can add mulitple relations of the same kind to a single source entity!
+let cart  = world.spawn().id();
+commands.entity(alice).insert_relation(FriendsWith, cart); // Hi!
+```
+
+Once your relations are in place, you'll want to access them in some form.
+Just like with ordinary components, you can request them in your queries:
+
+```rust
+// Relations can be queried for immutably
+fn friendship_is_magic(mut query: Query<(&mut Magic, &Relation<FriendsWith>), With<Pony>>) {
+    // Each entity can have their own friends
+    for (mut magic, friends) in query.iter_mut(){
+        // Relations return an iterator when unpacked
+        magic.power += friends.count();
+    }
+}
+
+// Or you can request mutable access to modify the relation's data
+fn interest(query: Query<&mut Relation<Owes>>){
+    for debts in query.iter(){
+        for (owed_to, amount_owed) in debts {
+            println!("we owe {} some money", owed_to);
+            amount_owed *= 1.05; 
+        }
+    }
+}
+
+// You can query for entities that target a specific entity
+fn friend_of_dorothy(
+    mut query: Query<Entity, With<Relation<FriendsWith>>>, 
+    dorothy: Res<Dorothy>,
+) {
+    // Setting relation filters requires mutable access to the query
+    query.filter_relation::<FriendsWith, _>(
+        RelationFilter::target(dorothy.entity)
+    // .build() causes the relation filters to be applied
+    ).build();
```

Comment:
Why is builder pattern necessary here? If we do that, then the builder should return something.

```rust
let mut filtered_query = query.filter_relation(..).build();
```

Otherwise, adding more filters would be possible after building. And also querying things without calling `build` would lead to bugs, as the filters would be not applied.

Reply:
@BoxyUwU was explaining to me that the builder pattern was a necessary API sacrifice for performance in order to... something about archetype iteration? I'll add a to-do in rationale.

I think I agree with the comments about the return value though; I'll make that change.

Reply:
Oh hey, using assignment allows us to avoid needing mutable access *and* eliminates the need for a reset_filter API.

I'm in favor of this.